python/ray/util/sgd/v2/backends/backend.py

```diff
@@ -372,37 +383,37 @@ def end_training():
                 self._process_checkpoint(results)
 
         futures = self.worker_group.execute_async(end_training)
-        return self.get_with_failure_handling(futures)
+        results = self.get_with_failure_handling(futures)
+        return results
 
     def get_with_failure_handling(self, remote_values):
         """Gets the remote values while handling for worker failures.
 
+        This method should be called instead of ``ray.get()`` directly in
+        order to handle worker failures.
+
+        If a worker failure is identified, backend specific failure handling
+        is executed and a ``TrainingWorkerError`` is raised.
+
         Args:
             remote_values (list): List of object refs representing functions
                 that may fail in the middle of execution. For example, running
                 a SGD training loop in multiple parallel actor calls.
-
         Returns:
             The resolved objects represented by the passed in ObjectRefs.
         """
-        unfinished = remote_values
-        try:
-            while len(unfinished) > 0:
-                finished, unfinished = ray.wait(unfinished)
-                # If a failure occurs the ObjectRef will be marked as finished.
-                # Calling ray.get will expose the failure as a RayActorError.
-                ray.get(finished)
-        except RayActorError as exc:
-            logger.exception(str(exc))
-            self.handle_failure()
-            return
-        return ray.get(remote_values)
-
-    def handle_failure(self):
-        # TODO: Fault-tolerance/elastic training here.
-        self.shutdown()
-        raise RuntimeError("Worker crashed during training. "
-                           "Training unsuccessful.")
+        success, failed_worker_indexes = check_for_failure(remote_values)
+        if success:
+            return ray.get(remote_values)
+        else:
+            self._num_failures += 1
+            try:
+                self._backend.handle_failure(self.worker_group,
+                                             failed_worker_indexes,
+                                             self._backend_config)
+            except RayActorError:
+                self._restart()
+            raise TrainingWorkerError
```

Comment:
can you raise with some intelligible error message? Maybe the string of the Ray Actor Error?

Reply:
The user should never be exposed to `TrainingWorkerError`. Instead I changed it so that we log the error message instead of passing it through the `TrainingWorkerError`.

Reply:
The intention of the comment is more for internal debugging purposes

Reply:
The error message is being printed out anyways here https://github.com/ray-project/ray/pull/18090/files#diff-719dd761cccd53aaba210f04b4e509ed7ef14f63862c0e74cd12379ef1e7e2bfR35, so for internal debugging purposes, this should suffice.

It'll be difficult to actually propagate the message through to the `TrainingWorkerError` this since the `RayActorError` is being raised in a different function: https://github.com/ray-project/ray/pull/18090/files#diff-719dd761cccd53aaba210f04b4e509ed7ef14f63862c0e74cd12379ef1e7e2bfR34